fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -1219,6 +1229,51 @@ private ScoredPlan planRankWithAnd(@Nonnull CandidateScan candidateScan,
         return null;
     }
 
+    private ScanComparisons getScanForAndLucene(@Nonnull AndComponent filter, @Nullable FilterSatisfiedMask filterMask) {
+        final Iterator<FilterSatisfiedMask> subFilterMasks = filterMask != null ? filterMask.getChildren().iterator() : null;
+        final List<QueryComponent> filters = filter.getChildren();
+        ScanComparisons scanComparisons = ScanComparisons.EMPTY;
+        for (QueryComponent subFilter : filters) {
+            final FilterSatisfiedMask childMask = subFilterMasks != null ? subFilterMasks.next() : null;
+            ScanComparisons children = getComparisonsForLuceneFilter(subFilter, childMask);
+            if (children != null) {
+                childMask.setSatisfied(true);
+                return children;
+            }
+        }
+        return scanComparisons;
+    }
+
+    private ScanComparisons getComparisonsForLuceneFilter(@Nonnull QueryComponent filter, FilterSatisfiedMask filterMask) {
+        if (filter instanceof AndComponent) {
+            return getScanForAndLucene((AndComponent) filter, filterMask);
+        } else if (filter instanceof LuceneQueryComponent) {
+            return ScanComparisons.from(((LuceneQueryComponent)filter).getComparison());
+        }
+        return null;
+    }
+
+
+    @Nullable
+    private ScoredPlan planLucene(@Nonnull CandidateScan candidateScan,
+                                @Nonnull Index index, @Nonnull QueryComponent filter,
+                                @Nullable KeyExpression sort) {
+        if (sort != null) {
+            // TODO: Full Text: Sorts are not supported with full text queries (https://github.com/FoundationDB/fdb-record-layer/issues/55)
+            return null;
+        }
+
+        FilterSatisfiedMask filterMask = FilterSatisfiedMask.of(filter);
+        final ScanComparisons scans = getComparisonsForLuceneFilter(filter, filterMask);
+        if (scans == null) {
+            return null;
+        }
+        RecordQueryPlan plan;
+        plan = new RecordQueryIndexPlan(index.getName(), IndexScanType.BY_LUCENE, scans, false);
```

Comment:
Is `scans` really in a form that `BY_LUCENE` can digest? Specifically when there is more than one.

Would it be better to assemble various `LuceneQueryComponent`s into a single Lucene query using string operations?

Reply:
I think what this special sub-planner does is take a Boolean tree of:
* Lucene components with a field name (field has to be one of the ones in the Lucene index, either in the stored primary key fields or the unstored index only fields)
* Lucene query fragments without their own field name
* Regular query comparisons, which might or might not involve a field present in the index

From these it produces a _single_ Lucene query string, that drives the scan (Lucene `search`) and an optional residual query component that is checked at runtime against records returned by the Lucene cursor, that is, turned into a regular filter plan with Lucene plan inside.

Reply:
So in a previous test where I tested that precisely yes it actually worked. However the name scans is a bit misleading as to how the code actually works here. It works by returning one of the comparisons in the and component, similar to the textPlanner and then the record layer handles the unsatisfied filters in other logic. To your point about combining the queries into a single Lucene query, yeah, that makes more sense as to how to handle this case.

Reply:
Yes, except that the leftover was a Lucene query component. I am still unsure whether those will work at runtime or it is a requirement to have them all folded into the scan's search string.

Reply:
I'm not sure I'm understanding your concern. Both when I put the Scan Comparisons as separate scan comparisons into the same plan and when I leave the second Lucene filter as a unsatisfied filter the plan works and returns the correct documents when executed. The question though that we're trying to answer here I think is which is better and I think that is the third option that you mentioned above of combining the scans into the Lucene search string.

Reply:
Is there something that can take a Lucene query component and a record and tell whether the record matches?

Reply:
I think it was decided that since it works both ways implementing the flattening of multiple Lucene components into a complex query string should be postponed. We would also want to allow the planner to compare that scan with a scan that doesn't flatten them.

Reply:
I am okay with postponing, since the caller can do the flattening. But I still don't see how leftovers are evaluated for arbitrary Lucene expressions and not just simple comparisons.